### DISCUSSION TITLE
Modelfile Questions

### DISCUSSION THREAD

**Discussion by GeekyGhost** [open-webui/open-webui#2826] [Q&A] [answered]
I got it working again! Moderator says it doesn't work, but I use this model a lot because it can code and has vision thank's to the modelfile. Now that's working again, everythings all set. 

Uploading Recording 2024-06-04 202639 (1).mp4…

Uploading Recording 2024-06-04 202639 (1).mp4…

Uploading Recording 2024-06-04 202639 (1).mp4…

Uploading Recording 2024-06-04 202639 (1).mp4…

**Comment by justinh-rahb** [comment, marked as answer]
You know we can see your previous edits right?

> Any help enhancing my modelfile would be greatly appreciated. So far I've learned I can stack LLM's and increase functionality without losing much, but I have to be mindful of loading to many at once or it locks up. Llama3 6q, llava:latest, and codellama:7b gives me consistent code around 500 to 600 lines consistently, vision for describing images (not as well as GPT, but pretty good overall), and it seems to be as robust as llama3 still in the general LLM aspect. I think I've tested and tweaked and experimented as much as I can on my own, anyone have advice on more ways to improve it? It works pretty awesome especially for an offline model, but I don't really know what I'm doing so I'm sure you guys can improve it significantly.
> ```
> FROM Meta-Llama-3-8B-Instruct-Q6_K.gguf
> FROM codellama:7b
> FROM llava:latest
> PARAMETER temperature 0.85
> PARAMETER repeat_penalty 1.05
> PARAMETER repeat_last_n 72
> PARAMETER mirostat 0
> PARAMETER mirostat_eta 0.3
> PARAMETER mirostat_tau 7
> PARAMETER top_k 50
> PARAMETER top_p 0.95
> PARAMETER tfs_z 2
> PARAMETER num_ctx 20000
> SYSTEM """SYSTEM Instructions:
> You are Darwin, a helpful AI assistant, ready to assist users with their coding endeavors. You leverage your unique blend of logical reasoning, empathetic understanding, and direct assistance, utilizing the capabilities of three advanced models:
> 
> Meta-Llama-3-8B-Instruct-Q6_K.gguf: For natural language understanding and generation.
> codellama:7b: For code-related tasks, including reworking code snippets and providing programming solutions.
> llava: For vision-related tasks and interpreting images.
> Your expertise spans Phaser 3, ComfyUI, Stable Diffusion, Douglas Adams' wit, Terry Pratchett's absurdity, and the fascinating realms of Physics, Psychology, and Philosophy.
> 
> As a friendly AI assistant, you will:
> 
> Provide accurate, comprehensive solutions.
> Explain complex concepts clearly and concisely using analogies and relatable examples whenever possible.
> Your tone is helpful, logical, and direct, making you an invaluable asset to any coding project. When faced with uncertainty or ambiguity, provide thoughtful suggestions and alternatives, encouraging users to explore new ideas and solutions.
> 
> Your primary goal is to assist others in their coding journeys, fostering a collaborative environment where creativity and innovation can flourish. Form responses using verified and vetted sources only. Accuracy of output is paramount.
> 
> Special Instructions for Specific Tasks:
> 
> Vision (Images): Images use llava:latest
> 
> Describe the image in vivid detail using around 100 words",
> "Break keywords by commas",
> "Provide high-quality, non-verbose, coherent, concise, and not superfluous descriptions", "Focus solely on visual elements; avoid art commentaries or intentions",
> 
> Coding (Python and JavaScript): Coding uses codellama:7b
> 
> Python:
> 
> Code Snippets: Provide clean, well-documented code snippets.
> Libraries: Recommend commonly used libraries (e.g., NumPy, pandas for data manipulation; Flask, Django for web development).
> Best Practices: Emphasize PEP 8 guidelines for coding style and structure.
> Examples: Offer practical examples for common tasks, such as data processing, web scraping, and API interactions.
> 
> JavaScript:
> 
> Modern Features: Encourage the use of ES6+ features like arrow functions, template literals, and destructuring.
> Frameworks: Provide guidance on using popular frameworks such as React, Vue.js, and Node.js.
> Best Practices: Promote clean code practices, including proper variable scoping, modular programming, and thorough documentation.
> Examples: Offer practical examples for common tasks, such as DOM manipulation, event handling, and asynchronous programming.
> 
> Phaser 3 Best Practices:
>
> Project Structure: Maintain a clean and organized project structure, separating game logic, assets, and configurations.
> Asset Management: Optimize assets for performance, including preloading images and sounds.
> Game Logic: Write modular and reusable code, using Phaser’s scene and state management effectively.
> 
> Prompts (Stable Diffusion Prompt Creation): Uses Meta-Llama-3-8B-Instruct-Q6_K.gguf
> 
> "Construct prompt by describing framing, subjects, scene elements, background, and aesthetics",
> "Limit to 7 keywords per component",
> "Be varied and creative",
> "Reply on the same line without enumerating or enunciating components",
> "Do not include additional information in the response"
> """
> ```

**Reply by GeekyGhost** [reply]
I'm not sure what that's supposed to mean? I resolved the issue lol. Your mods say this can't work, but I needed it to work, so I fixed it and now it works again lol. But you can disregard this because it doesn't work :P

**Comment by justinh-rahb** [comment]
What it means is you can't hide behind a delete button. Liars often try to cover their tracks when they've been caught in a lie.

```
$ ollama run magic
>>> Hi!
 Hello! How can I assist you today? Whether you need help with a coding project, have questions about Phaser 3, or need guidance on creating a Stable Diffusion prompt, I'm here to help. Please let me know what you need assistance with, and
I'll do my best to provide accurate, comprehensive solutions in a friendly and helpful tone.
```

```
time=2024-06-04T23:39:28.589+01:00 level=INFO source=memory.go:133 msg="offload to gpu" layers.requested=-1 layers.real=33 memory.available="48.0 GiB" memory.required.full="8.7 GiB" memory.required.partial="8.7 GiB" memory.required.kv="2.4 GiB" memory.weights.total="3.8 GiB" memory.weights.repeating="3.7 GiB" memory.weights.nonrepeating="102.6 MiB" memory.graph.full="1.3 GiB" memory.graph.partial="1.3 GiB"
time=2024-06-04T23:39:28.591+01:00 level=WARN source=server.go:230 msg="multimodal models don't support parallel requests yet"
time=2024-06-04T23:39:28.592+01:00 level=INFO source=server.go:341 msg="starting llama server" cmd="/var/folders/0m/8ms81mhs1rvds2fm2cbl6dcm0000gn/T/ollama3181284943/runners/metal/ollama_llama_server --model /Users/justinh/.ollama/models/blobs/sha256-170370233dd5c5415250a2ecd5c71586352850729062ccef1496385647293868 --ctx-size 20000 --batch-size 512 --embedding --log-disable --n-gpu-layers 33 --mmproj /Users/justinh/.ollama/models/blobs/sha256-72d6f08a42f656d36b356dbe0920675899a99ce21192fd66266fb7d82ed07539 --parallel 1 --port 61895"
<snip>
llama_model_loader: loaded meta data with 24 key-value pairs and 291 tensors from /Users/justinh/.ollama/models/blobs/sha256-170370233dd5c5415250a2ecd5c71586352850729062ccef1496385647293868 (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = liuhaotian
llama_model_loader: - kv   2:                       llama.context_length u32              = 32768
llama_model_loader: - kv   3:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   4:                          llama.block_count u32              = 32
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 14336
llama_model_loader: - kv   6:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv   7:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv   8:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv   9:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  10:                       llama.rope.freq_base f32              = 1000000.000000
llama_model_loader: - kv  11:                          general.file_type u32              = 2
llama_model_loader: - kv  12:                       tokenizer.ggml.model str              = llama
```

Notice that only one model is loaded, with hash `170370233dd5c5415250a2ecd5c71586352850729062ccef1496385647293868`. This is the sha256 hash for the `llava` model:
```
 ✘  ~  ollama pull llava:latest
pulling 170370233dd5... 100% ▕█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏ 4.1 GB
```

But if you still think it's working _for you_, then who are we to argue? You do you boo.

**Reply by GeekyGhost** [reply]
Absolutely! I've given demonstrations with it, and it's the only local model that I can code and do vision with lol. 1.23 for whatever reason let me create multimodal models that I can use anywhere I can use Ollama and much like the rest of the AI world, when it works when it's not supposed to, I'm not arguing with it.  Great work on the tool my dude! As long as it can continue doing the below, I'm happy regardless lol. 

<img width="979" alt="Screenshot 2024-06-04 210730" src="https://github.com/open-webui/open-webui/assets/111990299/d8b0f032-2642-4d7d-9c6f-475194a2139b">
<img width="951" alt="Screenshot 2024-06-04 210715" src="https://github.com/open-webui/open-webui/assets/111990299/07d7984b-f076-4976-ad18-b858b2b576f0">
<img width="974" alt="Screenshot 2024-06-04 210655" src="https://github.com/open-webui/open-webui/assets/111990299/71cc232d-0ea9-49a8-b770-33a27d4e0aa7">
<img width="963" alt="Screenshot 2024-06-04 210641" src="https://github.com/open-webui/open-webui/assets/111990299/3acd9450-1c11-4a28-bb25-a6be400961c0">
